okio/jvm/src/main/java/okio/Pipe.kt

```diff
@@ -66,25 +74,81 @@ class Pipe(internal val maxBufferSize: Long) {
           (buffer as Object).notifyAll() // Notify the source that it can resume reading.
         }
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { write(source, byteCount) } }
     }
 
     override fun flush() {
+      var delegate: Sink? = null
       synchronized(buffer) {
         check(!sinkClosed) { "closed" }
-        if (sourceClosed && buffer.size > 0L) throw IOException("source is closed")
+
+        foldedSink?.let {
+          delegate = it
+          return@synchronized
+        }
+
+        if (sourceClosed && buffer.size > 0L) {
+          throw IOException("source is closed")
+        }
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { flush() } }
     }
 
     override fun close() {
+      var delegate: Sink? = null
       synchronized(buffer) {
         if (sinkClosed) return
+
+        foldedSink?.let {
+          delegate = it
+          return@synchronized
+        }
+
         if (sourceClosed && buffer.size > 0L) throw IOException("source is closed")
         sinkClosed = true
         (buffer as Object).notifyAll() // Notify the source that no more bytes are coming.
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { close() } }
     }
 
     override fun timeout(): Timeout = timeout
+
+    private fun honorsTimeoutAndDeadline(delegateSink: Sink, block: Sink.() -> Unit) {
```

Comment:
There’s a very constrained way you always invoke this:

```
delegate?.let { honorsTimeoutAndDeadline(it) { ... } }
```

Here’s a proposal for another signature

```
fun Sink.forward(block: Sink.() -> Unit)
```

And then the callsite could be:

```
delegate?.forward { ... }
```

Reply:
had to move it out of the Sink class then otherwise would not be able to have access to `this@Pipe.sink`